File: scraper/Course.py
from typing import List, Tuple, Dict, DefaultDict, Union, Optional
from collections import defaultdict
import math
import json
import re
import os
from Stream import Stream

__all__ = ['Course']

os.chdir(os.path.dirname(os.path.abspath(__file__)))

with open('courses.json') as f:
    COURSE_NAMES = json.load(f)


class Course:
    def __init__(self, code: str, name: str, streams: List[Stream]) -> None:
        self.code = code
        self.name = name
        self.streams = streams

        # Get term disambiguation
        self.term: Optional[str] = None
        termMatch = re.search(r' \(([A-Z][A-Z0-9]{2})\)', name)
        if termMatch is not None:
            self.term = termMatch.group(1)

    @property
    def full_name(self):
        if self.code in COURSE_NAMES:
            return COURSE_NAMES[self.code]

        # If no full name is found, fall back to the initial (abridged) name
        return self.name

    def add_stream(self, component, section, status, enrols, times):
        # Create stream object
        stream = Stream(
            component=component,
            section=section,
            status=status,
            enrols=enrols,
            times=times
        )

        # Include the stream iff all it's attributes are valid
        if stream.included:
            self.streams.append(stream)

        return stream

    def remove_duplicates(self):
        # Group streams into sets by component and time
        stream_sets: DefaultDict[Tuple[str, Tuple[str]], List[Stream]] = defaultdict(list)
        for stream in self.streams:
            # Get stream times
            if not stream.web:
                # Extract just the times
                # (there is also location and weeks data in the tuples that we
                #  don't want to group on)
                times = tuple(time[0] for time in stream.times)
            else:
                # Web streams don't have any associated time
                times = None

            stream_sets[(stream.component, times)].append(stream)

        # Remove all duplicate streams (same course, component, and time)
        to_delete = []
        for streams in stream_sets.values():
            # Get the stream with the emptiest class (emptiest by ratio)
            best_stream = None
            best_ratio = math.inf
            for stream in streams:
                enrols, capacity = stream.enrols.split('/')
                ratio = float(enrols) / float(capacity)

                if ratio < best_ratio:
                    best_ratio = ratio
                    best_stream = stream

            # Remove every other stream
            to_delete.extend(stream for stream in streams if stream is not best_stream)

        # Remove streams marked for deletion
        for stream in to_delete:
            self.streams.remove(stream)

    def to_dict(self) -> Dict[str, Union[str, List[Dict[str, str]]]]:
        result = {
            'c': self.code,
            'n': self.full_name,
            's': [stream.to_dict() for stream in self.streams]
        }

        if self.term is not None:
            result['t'] = self.term

        return result

File: src/state_old/Course.ts
import { Stream, RawStreamData, StreamData } from './Stream';
const CourseNames: { [code: string]: string } = require('../../src/assets/courses.json');

export const CBS_CODE = 'CBS';

export type CourseId = string;

export interface RawCourseData {
  code: string;
  name: string;
}

export interface CourseData {
  code: string;
  name: string;
  streams: StreamData[];
  term?: string | null;
  isCustom?: boolean;
}

export class Course {
  code: string;
  name: string;
  streams: Stream[];
  term: string | null;
  isCustom: boolean;

  constructor(courseData: CourseData) {
    this.code = courseData.code;
    this.name = courseData.name;
    this.streams = courseData.streams.map(s => new Stream({ ...s, course: this }));
    this.term = courseData.term || null;
    this.isCustom = courseData.isCustom || false;
  }

  static from (data: RawCourseData) {
    const code = data.code.trim();
    const term = (/ \(([A-Z][A-Z0-9]{2})\)/.exec(data.name) || [])[1];
    const fullName = CourseNames[code];
    const name = fullName || data.name.trim().replace(new RegExp(`\\s*\\(${term}\\)$`), '');
    return new Course({
      code,
      name,
      streams: [],
      term,
    });
  }

  get data (): CourseData {
    return {
      code: this.code,
      name: this.name,
      streams: this.streams.map(s => s.data),
      term: this.term,
      isCustom: this.isCustom,
    }
  }

  get id (): CourseId {
    return this.code + (this.term || '');
  }

  addStream (data: RawStreamData) {
    const stream = Stream.from(data, this);
    if (stream !== null) {
      this.streams.push(stream);
    }
  }

  updateWith (data: Partial<CourseData>) {
    return new Course({
      ...this.data,
      ...data,
    });
  }

  removeDuplicates () {
    const mapping = new Map<string, Stream[]>();
    for (let stream of this.streams) {
      const times = stream.times !== null ? stream.times.map(t => t.time) : null;
      const key = stream.component + `[${times}]`;
      mapping.set(key, (mapping.get(key) || []).concat(stream));
    }

    // For each set of streams with identical component and times,
    for (const streams of Array.from(mapping.values())) {
      const emptiest = emptiestStream(streams);
      for (let stream of streams) {
        if (stream !== emptiest) {
          this.streams.splice(this.streams.indexOf(stream), 1);
        }
      }
    }
  }

  get hasWebStream () {
    return this.streams.filter(stream => stream.web).length > 0;
  }
}

const emptiestStream = (streams: Stream[]) => {
  let bestStream = null;
  let bestRatio = Infinity;
  for (let stream of streams) {
    const ratio = stream.enrols[0] / stream.enrols[1];
    if (ratio < bestRatio) {
      bestRatio = ratio;
      bestStream = stream;
    }
  }

  return bestStream!;
}

export default Course;
